pkg/controllers/inventory/service_handler.go

```diff
@@ -0,0 +1,160 @@
+			log.Error(err, "DeletedFinalStateUnknown Obj is not *v1.Service")
+			return
+		}
+	}
+	log.V(1).Info("Inventory processing Service", "Service", service.Name, "Namespace", service.Namespace)
+	key, _ := keyFunc(service)
+	log.V(1).Info("Adding Service key to inventory object queue", "Service key", key)
+	c.inventoryObjectQueue.Add(inventory.InventoryKey{InventoryType: inventory.ContainerApplication, ExternalId: string(service.UID), Key: key})
+}
+
+func watchEndpoint(c *InventoryController, mgr ctrl.Manager) error {
+	endpointInformer, err := mgr.GetCache().GetInformer(context.Background(), &v1.Endpoints{})
+	if err != nil {
+		log.Error(err, "Failed to create Endpoints informer")
+		return err
+	}
+
+	_, err = endpointInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.handleEndpoint(obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			c.handleEndpoint(newObj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.handleEndpoint(obj)
+		},
+	})
+	if err != nil {
+		log.Error(err, "Failed to add Endpoints event handler")
+		return err
+	}
+	return nil
+}
+
+func (c *InventoryController) handleEndpoint(obj interface{}) {
+	var endpoint *v1.Endpoints
+	ok := false
+	switch obj1 := obj.(type) {
+	case *v1.Endpoints:
+		endpoint = obj1
+	case cache.DeletedFinalStateUnknown:
+		endpoint, ok = obj1.Obj.(*v1.Endpoints)
+		if !ok {
+			err := fmt.Errorf("obj is not valid *v1.Endpoints")
+			log.Error(err, "DeletedFinalStateUnknown Obj is not *v1.Endpoints")
+			return
+		}
+	}
+	log.V(1).Info("Inventory processing Endpoint", "Endpoint", endpoint.Name, "Namespace", endpoint.Namespace)
+	key, _ := keyFunc(endpoint)
+	service := &v1.Service{}
+	err := c.Client.Get(
+		context.TODO(),
+		types.NamespacedName{
+			Name:      endpoint.Name,
+			Namespace: endpoint.Namespace,
+		},
+		service,
+	)
+	if err != nil {
+		log.Error(err, "Failed to get Service", "Name", endpoint.Name, "Namespace", endpoint.Namespace)
+		return
+	}
+
+	// Get endpoint's corresponding pods and enqueue them
+	podIds, _ := c.service.GetPodIDsFromEndpoint(service)
+	for _, podID := range podIds {
+		pod, err := GetPodByUID(context.TODO(), c.Client, types.UID(podID), endpoint.Namespace)
+		if err != nil {
+			log.Error(err, "Failed to get Pod by UID", "PodID", podID, "Namespace", endpoint.Namespace)
+			continue
+		}
+		podKey, keyErr := keyFunc(pod)
+		if keyErr != nil {
+			log.Error(keyErr, "Failed to get key for Pod", "Pod", pod.Name, "Namespace", pod.Namespace)
+			continue
+		}
+		log.V(1).Info("Adding Endpoint ContainerApplicationInstance key to inventory object queue", "ContainerApplicationInstance ExternalId", podID, "ContainerApplicationInstance Key", podKey)
+		// Ensure pod builder can successfully get ContainerApplicationIds
+		containerApplicationInstance := containerinventory.ContainerApplicationInstance{
+			DisplayName:             pod.Name,
+			ContainerApplicationIds: []string{string(service.UID)},
```

Comment:
can't assume only one service select the pod

Reply:
fixed